File: typescript/src/models/MediaReaction.ts
export enum MediaReactionType {
  Like       = 1,
  Share      = 2,
  Comment    = 3,
  SuperReact = 4,
}

export interface MediaReaction {
  reactionId: string;      // UUID string
  contentHash: string;
  fromUhid: string;
  type: MediaReactionType;
  positionMs: number;
  message: string | null;  // required for Comment, null otherwise
  sentAtMs: number;        // unix milliseconds
}

// ── Wire format ───────────────────────────────────────────────────────────────

/** Snake_case JSON representation of MediaReaction as sent/received on the wire. */
export interface MediaReactionWire {
  reaction_id: string;
  content_hash: string;
  from_uhid: string;
  type: string;            // lowercase enum string: "like" | "share" | "comment" | "super_react"
  position_ms: number;
  message: string | null;
  sent_at_ms: number;
}

const reactionTypeToWire: Record<MediaReactionType, string> = {
  [MediaReactionType.Like]:       "like",
  [MediaReactionType.Share]:      "share",
  [MediaReactionType.Comment]:    "comment",
  [MediaReactionType.SuperReact]: "super_react",
};

const wireToReactionType: Record<string, MediaReactionType> = {
  like:        MediaReactionType.Like,
  share:       MediaReactionType.Share,
  comment:     MediaReactionType.Comment,
  super_react: MediaReactionType.SuperReact,
};

export function toWire(reaction: MediaReaction): MediaReactionWire {
  return {
    reaction_id:  reaction.reactionId,
    content_hash: reaction.contentHash,
    from_uhid:    reaction.fromUhid,
    type:         reactionTypeToWire[reaction.type],
    position_ms:  reaction.positionMs,
    message:      reaction.message,
    sent_at_ms:   reaction.sentAtMs,
  };
}

export function fromWire(obj: MediaReactionWire): MediaReaction {
  const type = wireToReactionType[obj.type];
  if (type === undefined) throw new Error(`Unknown reaction type: ${obj.type}`);
  return {
    reactionId:  obj.reaction_id,
    contentHash: obj.content_hash,
    fromUhid:    obj.from_uhid,
    type,
    positionMs:  obj.position_ms,
    message:     obj.message,
    sentAtMs:    obj.sent_at_ms,
  };
}

// ── Factory ───────────────────────────────────────────────────────────────────

/**
 * Factory that validates the reaction rules before constructing the object.
 *
 * - Comment requires a non-empty message.
 * - All other types must have message === null.
 */
export function createReaction(
  reactionId: string,
  contentHash: string,
  fromUhid: string,
  type: MediaReactionType,
  positionMs: number,
  message: string | null,
  sentAtMs: number,
): MediaReaction {
  if (!contentHash.trim()) throw new Error("contentHash must not be empty");
  if (!fromUhid.trim())   throw new Error("fromUhid must not be empty");
  if (positionMs < 0)     throw new Error("positionMs must be >= 0");

  if (type === MediaReactionType.Comment) {
    if (!message || !message.trim()) {
      throw new Error("A message is required for Comment reactions");
    }
  } else {
    if (message !== null) {
      throw new Error(`message must be null for ${MediaReactionType[type]} reactions`);
    }
  }

  return { reactionId, contentHash, fromUhid, type, positionMs, message, sentAtMs };
}

File: src/AetherMesh.Media.Core/Models/MediaReaction.cs
using System.Text.Json.Serialization;

namespace AetherMesh.Media.Core.Models;

/// <summary>
/// A timestamped reaction sent by a viewer to the creator of a piece of content.
/// </summary>
public sealed record MediaReaction
{
    /// <summary>Unique identifier for this reaction event.</summary>
    [JsonPropertyName("reaction_id")]
    public Guid ReactionId { get; init; }

    /// <summary>SHA-256 hex hash of the content being reacted to.</summary>
    [JsonPropertyName("content_hash")]
    public string ContentHash { get; init; }

    /// <summary>Uhid of the viewer sending the reaction.</summary>
    [JsonPropertyName("from_uhid")]
    public string FromUhid { get; init; }

    /// <summary>The kind of reaction.</summary>
    [JsonPropertyName("type")]
    public MediaReactionType Type { get; init; }

    /// <summary>
    /// Position within the media (in milliseconds) where the reaction occurred.
    /// A value of 0 means the reaction was not anchored to a specific position.
    /// </summary>
    [JsonPropertyName("position_ms")]
    public long PositionMs { get; init; }

    /// <summary>
    /// Free-text message body.  Required when <see cref="Type"/> is
    /// <see cref="MediaReactionType.Comment"/>; must be <c>null</c> for all
    /// other reaction types.
    /// </summary>
    [JsonPropertyName("message")]
    public string? Message { get; init; }

    /// <summary>Wall-clock unix milliseconds at which the reaction was sent.</summary>
    [JsonPropertyName("sent_at_ms")]
    public long SentAtMs { get; init; }

    /// <summary>
    /// Initialises and validates a <see cref="MediaReaction"/>.
    /// </summary>
    /// <param name="reactionId">Unique reaction identifier.</param>
    /// <param name="contentHash">SHA-256 hex of the target content.</param>
    /// <param name="fromUhid">Sender's Uhid.</param>
    /// <param name="type">Reaction type.</param>
    /// <param name="positionMs">Playback position in milliseconds.</param>
    /// <param name="message">Comment text (only for <see cref="MediaReactionType.Comment"/>).</param>
    /// <param name="sentAtMs">Unix millisecond timestamp of the reaction.</param>
    /// <exception cref="ArgumentException">
    /// Thrown when a <see cref="MediaReactionType.Comment"/> is missing its
    /// message, or when a non-comment reaction carries a message.
    /// </exception>
    public MediaReaction(
        Guid reactionId,
        string contentHash,
        string fromUhid,
        MediaReactionType type,
        long positionMs,
        string? message,
        long sentAtMs)
    {
        if (string.IsNullOrWhiteSpace(contentHash))
            throw new ArgumentException("ContentHash must not be empty.", nameof(contentHash));
        if (string.IsNullOrWhiteSpace(fromUhid))
            throw new ArgumentException("FromUhid must not be empty.", nameof(fromUhid));
        if (positionMs < 0)
            throw new ArgumentOutOfRangeException(nameof(positionMs), "PositionMs must be >= 0.");

        if (type == MediaReactionType.Comment)
        {
            if (string.IsNullOrWhiteSpace(message))
                throw new ArgumentException(
                    "A Message is required for Comment reactions.", nameof(message));
        }
        else
        {
            if (message is not null)
                throw new ArgumentException(
                    $"Message must be null for {type} reactions.", nameof(message));
        }

        ReactionId  = reactionId;
        ContentHash = contentHash;
        FromUhid    = fromUhid;
        Type        = type;
        PositionMs  = positionMs;
        Message     = message;
        SentAtMs    = sentAtMs;
    }
}
